### DISCUSSION TITLE
.env in production

### DISCUSSION THREAD

**Discussion by iammrsea** [adonisjs/core#1353] [Help] [answered]
Has anyone successfully deployed V5 in production? I am having challenges deploying mine to heroku, especially with the .env file.
It appears that the EnvProvider relies on the existence of .env file (even in production) to load and parse the environment variables. I am deploying my code from a github repository, so obviously, my .env file is missing. I was hoping that the EnvProvider should rely on process.env for the environment variables instead of the actual .env file. I was tempted to get rid of the EnvProvider and manually parse the environment variables with dotenv package, but after doing that, my ace commands started breaking.  Any ideas on how I might get around this problem?

**Comment by thetutlage** [comment]
You just need to set `ENV_SILENT` environment variable and then EnvProvider will not complain about the missing `.env` file.

```sh
ENV_SILENT node build/server.js
```

**Reply by iammrsea** [reply]
Please could you be more specific on how I could set that? I tried  "start": "SET ENV_SILENT=true &&  node build/server.js" but it didn't work. Of course I also did "start": "ENV_SILENT node build/server.js" and got "ENV_SILENT not recognized command error".

**Comment by msrumon** [comment, marked as answer]
For **Windows**:
```bash
$ set ENV_SILENT = true
$ set APP_KEY = node ace generate:key
$ node build/server
```
Let me know if this works or not.

**Reply by iammrsea** [reply]
Thanks, I had it working already. I used heroku cli to set the variables.

**Reply by msrumon** [reply]
Good to know. Can you please share what steps you had taken to resolve the issue and mark that as answer? So that others can see it as reference.

**Comment by iammrsea** [comment]
**Steps to deploying V5 to heroku**

1. Set up your environment variables. You can do this through your dashboard on heroku or using heroku cli. The necessary environment variables are APP_KEY, ENV_SILENT and DB_CONNECTION (if you are using database). You can set others according to your requirements. For heroku cli, you can set the variables on your terminal as follows:
```javascript
$ heroku config:set APP_KEY=your_app_key_here --app=name_of_the_app_you_created_on_heroku;
$ heroku config:set ENV_SILENT=true --app=same_as_above;
$ heroku config:set DB_CONNECTION=db_driver (e.g 'pg' for postgres) --app=same_as_above
```
2. In your package.json file, make sure you have the following scripts:
```javascript
"build": "node ace build --production",
"start": "node build/server.js"
```
3. You can optionally create a Procfile in your root directory and add the content below, change yarn to npm if that's your preference
```javascript
web : yarn run start
```
Note, option 2 is a requirement if you are deploying directly from your github repo. 
That's all you need, of course you can add other things like postbuild, preinstall scripts to your package.json

**Reply by thiagolobao** [reply]
Hi. I follow those steps and my logs on Heroku show me this:
```
Process exited with status 0
State changed from starting to crashed
```

---
My Procfile
`web: yarn run start`

Can you help me?

Thx

**Reply by iammrsea** [reply]
Do you have heroku cli installed locally on your machine? If you do, could you run this command `heroku logs --tail --app=your_app_name`?Copy and paste the logs here, perhaps that could give a better picture of what caused the problem

**Reply by marcosrodriigues** [reply]
Hello,

I am with same problem here, and I tried what @iammrsea did but it still not working.

When I run `heroku logs --tail`, I can check my app can build and it seems that is running. When I try to access the url, the logs show me the follow:

2020-08-15T01:58:34.242358+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
2020-08-15T01:58:34.265209+00:00 heroku[web.1]: Stopping process with SIGKILL
2020-08-15T01:58:34.345269+00:00 heroku[web.1]: Process exited with status 137
2020-08-15T01:58:34.390985+00:00 heroku[web.1]: State changed from starting to crashed

I have set the PORT at my environment variables.

Could anyone can help?

**Reply by thiagolobao** [reply]
Hi @marcosrodriigues.

Heroku uses dynamic ports to server our applications.
I solved my problem removing `PORT` and `HOST` variables from my env file and add only `HOST` on heroku dashboard.
Try this.

Cheers.

**Reply by marcosrodriigues** [reply]
> Hi @marcosrodriigues.
> 
> Heroku uses dynamic ports to server our applications.
> I solved my problem removing `PORT` and `HOST` variables from my env file and add only `HOST` on heroku dashboard.
> Try this.
> 
> Cheers.

Hi @thiagolobao , thanks for your reply.

I solved the problem. I don't know exactly whats happen, but I have changed my Host from "localhost" to "0.0.0.0" on heroku dashboard and it's works!

Thanks at all!
